core/slice.go

```diff
@@ -694,29 +694,85 @@ func (sl *Slice) ConstructLocalBlock(header *types.Header) (*types.Block, error)
 		// in the subordinate manifest.
 		return types.NewBlockWithHeader(header), nil
 	}
-	pendingBlockBody := sl.PendingBlockBody(header.Root())
-	if pendingBlockBody == nil {
-		return nil, ErrBodyNotFound
-	}
-	// Load uncles because they are not included in the block response.
-	txs := make([]*types.Transaction, len(pendingBlockBody.Transactions))
-	for i, tx := range pendingBlockBody.Transactions {
-		txs[i] = tx
-	}
-	uncles := make([]*types.Header, len(pendingBlockBody.Uncles))
-	for i, uncle := range pendingBlockBody.Uncles {
-		uncles[i] = uncle
-		log.Debug("Pending Block uncle", "hash: ", uncle.Hash())
+	var block *types.Block
+	// check if the header has empty uncle and tx root
+	if header.EmptyBody() {
+		// construct block with empty transactions and uncles
+		block = types.NewBlockWithHeader(header)
+	} else {
+		pendingBlockBody := rawdb.ReadPendingBlockBody(sl.sliceDb, header.Hash())
+		if pendingBlockBody != nil {
+			// Load uncles because they are not included in the block response.
+			txs := make([]*types.Transaction, len(pendingBlockBody.Transactions))
+			for i, tx := range pendingBlockBody.Transactions {
+				txs[i] = tx
+			}
+
+			uncles := make([]*types.Header, len(pendingBlockBody.Uncles))
+			for i, uncle := range pendingBlockBody.Uncles {
+				uncles[i] = uncle
+				log.Debug("Pending Block uncle", "hash: ", uncle.Hash())
+			}
+
+			etxs := make([]*types.Transaction, len(pendingBlockBody.ExtTransactions))
+			for i, etx := range pendingBlockBody.ExtTransactions {
+				etxs[i] = etx
+			}
+
+			subBlockHashes := make(types.BlockManifest, len(pendingBlockBody.SubManifest))
+			for i, blockHash := range pendingBlockBody.SubManifest {
+				subBlockHashes[i] = blockHash
+			}
+
+			block = types.NewBlockWithHeader(header).WithBody(txs, uncles, etxs, subBlockHashes)
+		} else {
+			return nil, ErrBodyNotFound
+		}
 	}
-	etxs := make([]*types.Transaction, len(pendingBlockBody.ExtTransactions))
-	for i, etx := range pendingBlockBody.ExtTransactions {
-		etxs[i] = etx
+	if err := sl.validator.ValidateBody(block); err != nil {
+		return block, err
+	} else {
+		return block, nil
 	}
-	subManifest := make(types.BlockManifest, len(pendingBlockBody.SubManifest))
-	for i, blockHash := range pendingBlockBody.SubManifest {
-		subManifest[i] = blockHash
+}
+
+// constructLocalMinedBlock takes a header and construct the Block locally
+func (sl *Slice) ConstructLocalMinedBlock(header *types.Header) (*types.Block, error) {
+	var block *types.Block
+	// check if the header has empty uncle and tx root
+	if header.EmptyBody() {
+		// construct block with empty transactions and uncles
+		block = types.NewBlockWithHeader(header)
+	} else {
+		pendingBlockBody := sl.GetPendingBlockBody(header.Root())
```

Comment:
Shouldn't `ConstructLocalMinedBlock()` be looking in the worker's LRU cache or something like that? Why is it reading from the PendingBlockBody db?

Reply:
It is not reading from the db. GetPendingBlockBody reads from the worker cache